types/propEq.d.ts

```diff
@@ -1,6 +1,29 @@
-export function propEq<T>(val: T): {
-  <K extends PropertyKey>(name: K): (obj: Record<K, T>) => boolean;
-  <K extends PropertyKey>(name: K, obj: Record<K, T>): boolean;
+import { Placeholder } from 'ramda';
+
+export function propEq(__: Placeholder): never;
+export function propEq<V>(val: V): {
+  <K extends PropertyKey>(name: K): (obj: Record<K, any>) => boolean;
+  <U extends Record<PropertyKey, any>>(__: Placeholder, obj: U): (name: keyof U) => boolean;
+  <K extends keyof U, U extends Record<PropertyKey, any>>(name: K, obj: U): boolean;
 };
-export function propEq<T, K extends PropertyKey>(val: T, name: K): (obj: Record<K, T>) => boolean;
-export function propEq<K extends keyof U, U>(val: U[K], name: K, obj: U): boolean;
+export function propEq<K extends PropertyKey>(__: Placeholder, name: K): {
+  (__: Placeholder, obj: Record<K, any>): <V>(val: V) => boolean;
+  <V>(val: V, obj: Record<K, any>): boolean
+  <V>(val: V): (obj: Record<K, any>) => boolean
+};
+export function propEq<V, K extends PropertyKey>(val: V, name: K): (obj: Record<K, any>) => boolean;
+
+
+export function propEq<U extends Record<any, any>>(__: Placeholder, ___: Placeholder, obj: U): {
+  (__: Placeholder, name: keyof U): (__: Placeholder) => boolean;
+  (__: Placeholder, name: keyof U): <V>(val: V) => boolean;
+  <V>(val: V, name: keyof U): boolean;
+  (val: Placeholder): never;
+  <V>(val: V): (name: keyof U) => boolean;
```

Comment:
Per my main comment, I don't think `(val: Placeholder): never` is needed, since `val` should be `U[K]` here, and thus can never be `Placeholder`

Reply:
I tried to demonstrate in my main comment that this kind of typing is more likely to create more unnecessary obstacles, rather than help developers catch errors in the early stage